js-api-spec/compiler.test.ts

```diff
@@ -0,0 +1,140 @@
+// Copyright 2023 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import type {AsyncCompiler, Compiler, CompileResult} from 'sass';
+import {initAsyncCompiler, initCompiler, SassString} from 'sass';
+
+import {spy, URL} from './utils';
+
+export const functions = {
+  'foo($args)': (args: unknown) => new SassString(`${args}`),
+};
+
+export const importers = [
+  {
+    canonicalize: (url: string) => new URL(`u:${url}`),
+    load: (url: typeof URL) => ({
+      contents: `.import {value: ${url.pathname}} @debug "imported";`,
+      syntax: 'scss' as const,
+    }),
+  },
+];
+
+export const asyncImporters = [
+  {
+    canonicalize: (url: string) =>
+      Promise.resolve(importers[0].canonicalize(url)),
+    load: (url: typeof URL) => Promise.resolve(importers[0].load(url)),
+  },
+];
+
+export const getLogger = () => ({debug: spy(() => {})});
+
+/* A slow importer that executes a callback after a delay */
+export const getSlowImporter = (callback: () => void) => ({
+  canonicalize: async () => new URL('foo:bar'),
+  load: async () => {
+    await new Promise(resolve => setTimeout(resolve, 100));
```

Comment:
Rather than using a timeout when testing that concurrent systems wait for a given event, it's usually better to have a promise whose resolution you control explicitly so you can guarantee it won't finish until you want it to (and then you don't end up writing tests that take a lot longer than they need to because of delays). You can run `await Promise(resolve => setTimeout(resolve, 0))` to force all outstanding _non_-time-based tasks to complete, then resolve the promise and verify that the behavior happens as expected.

Reply:
Resolved in  6dcdd3824f575e495c25bd5e21d2d406453fe2f2